service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java

```diff
@@ -321,6 +322,13 @@ public Object run() throws HiveSQLException {
             LOG.error("Error running hive query: ", e);
           } finally {
             LogUtils.unregisterLoggingContext();
+            Hive hiveDb = Hive.getThreadLocal();
+            if (hiveDb != null && hiveDb != parentHive) {
```

Comment:
Another issue is parentHive is closed by the background thread if it is changed. This will cause problems if parentHive is accessed by caller after returning from BackgroundTask. I think we need not set parentHive in this task.